# Data Engineer Challenge Solutions

IMPORTING NECESSARY LIBRARIES

In [1]:
import pandas as pd
import numpy as np


READING CSV FILES

In [2]:
goalscorers_data = pd.read_csv('goalscorers.csv')
shootout_data = pd.read_csv('shootouts.csv')
results_data = pd.read_csv('results.csv')

GOALSCORERS DATAS

In [3]:
#goalscorers_data
goalscorers_df = pd.DataFrame(goalscorers_data)
goalscorers_data.info()
print('***********************************************************************')
print(goalscorers_data.describe())
print('***********************************************************************')
goalscorers_data.head()
#goalscorers_data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43189 entries, 0 to 43188
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date       43189 non-null  object 
 1   home_team  43189 non-null  object 
 2   away_team  43189 non-null  object 
 3   team       43189 non-null  object 
 4   scorer     43139 non-null  object 
 5   minute     42929 non-null  float64
 6   own_goal   43187 non-null  object 
 7   penalty    43187 non-null  object 
dtypes: float64(1), object(7)
memory usage: 2.6+ MB
***********************************************************************
             minute
count  42929.000000
mean      50.040206
std       26.363782
min        1.000000
25%       28.000000
50%       51.000000
75%       73.000000
max      122.000000
***********************************************************************


date  home_team away_team       team            scorer  minute  \
0  1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    44.0   
1  1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    55.0   
2  1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    70.0   
3  1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    75.0   
4  1916-07-06  Argentina     Chile  Argentina     Alberto Ohaco     2.0   

  own_goal penalty  
0    False   False  
1    False   False  
2    False   False  
3    False   False  
4    False   False

SHOOTOUT DATAS

In [4]:
#shootout_data
shootout_df = pd.DataFrame(shootout_data)
pd.set_option('display.max_columns', None)
shootout_df.info()
print('***********************************************************************')
print(shootout_df.describe())
print('***********************************************************************')
shootout_df.head()
# shootout_df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 562 entries, 0 to 561
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   date           562 non-null    object
 1   home_team      562 non-null    object
 2   away_team      562 non-null    object
 3   winner         562 non-null    object
 4   first_shooter  86 non-null     object
dtypes: object(5)
memory usage: 22.1+ KB
***********************************************************************
              date     home_team away_team     winner first_shooter
count          562           562       562        562            86
unique         498           165       178        163            29
top     2016-06-03  South Africa    Uganda  Argentina      Colombia
freq             5            15        15         14             9
***********************************************************************


date    home_team         away_team       winner first_shooter
0  1967-08-22        India            Taiwan       Taiwan           NaN
1  1971-11-14  South Korea  Vietnam Republic  South Korea           NaN
2  1972-05-07  South Korea              Iraq         Iraq           NaN
3  1972-05-17     Thailand       South Korea  South Korea           NaN
4  1972-05-19     Thailand          Cambodia     Thailand           NaN

RESULTS DATAS

In [5]:
#results_data
results_df = pd.DataFrame(results_data)
results_data.info()
print('***********************************************************************')
print(results_data.describe())
print('***********************************************************************')
results_data.head()
#results_data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45315 entries, 0 to 45314
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        45315 non-null  object
 1   home_team   45315 non-null  object
 2   away_team   45315 non-null  object
 3   home_score  45315 non-null  int64 
 4   away_score  45315 non-null  int64 
 5   tournament  45315 non-null  object
 6   city        45315 non-null  object
 7   country     45315 non-null  object
 8   neutral     45315 non-null  bool  
dtypes: bool(1), int64(2), object(6)
memory usage: 2.8+ MB
***********************************************************************
         home_score    away_score
count  45315.000000  45315.000000
mean       1.739314      1.178241
std        1.746904      1.392095
min        0.000000      0.000000
25%        1.000000      0.000000
50%        1.000000      1.000000
75%        2.000000      2.000000
max       31.000000     21.000000
***************

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  Scotland   England           0           0   Friendly  Glasgow   
1  1873-03-08   England  Scotland           4           2   Friendly   London   
2  1874-03-07  Scotland   England           2           1   Friendly  Glasgow   
3  1875-03-06   England  Scotland           2           2   Friendly   London   
4  1876-03-04  Scotland   England           3           0   Friendly  Glasgow   

    country  neutral  
0  Scotland    False  
1   England    False  
2  Scotland    False  
3   England    False  
4  Scotland    False

# FIRST TASK

Create a query that calculates the average number of goals per game between 1900 and 2000.



In [6]:
# Converting 'date' columns datas from string to date format
results_df['date'] = pd.to_datetime(results_df['date'])

# Filtering the 'results' dataframe to see the results between 1900-2000
results_1900to2000 = results_df[(results_df['date'] >= '1900-01-01') & (results_df['date'] < '2001-01-01')]
#results_1900to2000.head()

# Counting the number of the games between 1900-2000
number_of_games_1900to2000 = results_1900to2000.shape[0]
print('Number of the games between 1900 and 2000:',number_of_games_1900to2000)

# Calculating the total number of the goals scored between 1900 and 2000
total_goals_1900to2000 = results_1900to2000['home_score'].sum() + results_1900to2000['away_score'].sum()

# Calculating the average number of goals per game between 1900 and 2000
average_goals_1900to2000 = total_goals_1900to2000 / number_of_games_1900to2000
print("Average number of goals per game between 1900 and 2000 is =>",average_goals_1900to2000)

# Transforming the "date" type back to its string format
results_df['date'] = results_df['date'].dt.strftime('%Y-%m-%d')

Number of the games between 1900 and 2000: 23712
Average number of goals per game between 1900 and 2000 is => 3.0704284750337383


# SECOND TASK

Create a query that counts the number of shootouts wins by country and arrange in alphabetical order.



In [7]:
# Finding the number of the shootout wins by country
win_counts_by_country = shootout_df['winner'].value_counts()

# Arranging the number of the shootouts by country in alphabetical order
win_counts_by_country = win_counts_by_country.sort_index()

# Creating a dataframe from Series to make it more functional for future applications
win_counts_by_country_df = pd.DataFrame(win_counts_by_country)
win_counts_by_country_df
#win_counts_by_country_df

count
winner                    
Abkhazia                 2
Algeria                  7
Angola                   7
Antigua and Barbuda      2
Argentina               14
...                    ...
Zambia                  13
Zanzibar                 2
Zimbabwe                 7
Åland                    1
Åland Islands            2

[163 rows x 1 columns]

# THIRD TASK

Create a reliable key that allows the joining together of goal scorers, results, and shootouts.




In all of the three datasets "date", "home_team" and "away_team" columns are common. (and they have the same amount of rows)
So I created a key named "Game_ID" by using those three parameters. 
Which is gonna help me to use related datas from multiple datasets in the following tasks. 

In [8]:
goalscorers_df['Game_ID'] = goalscorers_df['date'] +'_'+ goalscorers_df['home_team'] +'_'+ goalscorers_df['away_team']

#goalscorers_df['Game_ID']

In [9]:
shootout_df['Game_ID'] = shootout_df['date'] +'_'+ shootout_df['home_team'] +'_'+ shootout_df['away_team']

In [10]:
results_df['Game_ID'] = results_df['date'] +'_'+ results_df['home_team'] +'_'+ results_df['away_team']


# FOURTH TASK



Create a query that identifies which teams have won a penalty shootout after a 1-1 draw.



In [11]:
# Finding the games ended with the result of " 1-1  " from the results dataframe
one_one_draw_games = results_df[(results_df['home_score']) == 1 & (results_df['away_score'] == 1)]
#number_of_one_one_draw_games = one_one_draw_games.count()
#print(number_of_one_one_draw_games)
#print(one_one_draw_games)
#one_one_draw_games_df = pd.DataFrame(one_one_draw_games)

# Adding a "Game_ID" column one_one_draw_games dataframe so we can define the games with unique id's
# used .loc[:,] to create 'Game_ID' for each column separately
one_one_draw_games.loc[:,'Game_ID'] = one_one_draw_games['date'] + '_' + one_one_draw_games['home_team'] + '_' + one_one_draw_games['away_team']

# Defining a subset of the shootout_df to avoid creating the same columns twice
shootout_df_subset = shootout_df[['Game_ID', 'winner']]
# Merging one_one_draw_games with shootout_df (on "Game_ID") and taking only the matched datas (The ones that ended 1-1 and went for penalty shootouts)
penalty_shootouts = one_one_draw_games.merge(shootout_df_subset, on='Game_ID', how='inner')


# Finding the teams have won a penalty shootout after a 1-1 draw
penalty_shootouts_winners = penalty_shootouts[['winner']]

penalty_shootouts_winners



winner
0         Taiwan
1    South Korea
2           Iraq
3    South Korea
4         Guinea
..           ...
415        India
416        India
417       Panama
418        Åland
419        Qatar

[420 rows x 1 columns]


# FIFTH TASK



Create a query that identifies the top goalscorer by tournament, and what percentage that equates to for all goals scored in the tournament.



1) Calculate the total number of the goals by tournament.

In [12]:
# Calculating the number of the goals for each game
# And adding a new column "total_goals" to the results dataframe
results_df["total_goals"] = results_df["home_score"] + results_df["away_score"]

# Grouping the data by 'tournament' column and summing up the total goals for each tournament
# after the grouping, tournament becomes an index so we need to reset it with reset_index() and make it a new column named 'total_goals'
tournament_goals = results_df.groupby('tournament')['total_goals'].sum().reset_index(name='total_goals')
#tournament_goals

2) Merge the goalscorers with results data so we can see the relationships between the goalscorer player and game results 

In [13]:
tournaments_and_goalscorers_df = goalscorers_df.merge(results_df, on='Game_ID', how='left')

3) Calculate the number of the goals for each player by tournament

In [14]:
# Grouping the data by 'tournament' and 'scorer'
# This groups the data based on each unique combination of tournament and scorer.
player_and_tournament = tournaments_and_goalscorers_df.groupby(['tournament','scorer']).size().reset_index(name='goal_count')

player_and_tournament

tournament                 scorer  goal_count
0            AFC Asian Cup           A'ala Hubail           5
1            AFC Asian Cup         Abbas Chahrour           1
2            AFC Asian Cup            Abbas Obeid           1
3            AFC Asian Cup             Abdah Alif           1
4            AFC Asian Cup      Abdelkarim Hassan           1
...                    ...                    ...         ...
19369  UEFA Nations League          Álvaro Morata           3
19370  UEFA Nations League            Ári Jónsson           1
19371  UEFA Nations League  Þórir Jóhann Helgason           2
19372  UEFA Nations League         İlkay Gündoğan           3
19373  UEFA Nations League          İsmail Yüksek           1

[19374 rows x 3 columns]

4) Calculate the maximum number of goals scored by one player for each tournament.

In [15]:
max_goals_tournament = player_and_tournament.groupby('tournament')['goal_count'].max().reset_index(name='max_goals')
max_goals_tournament

tournament  max_goals
0                    AFC Asian Cup         14
1           African Cup of Nations         18
2                       Baltic Cup          9
3        British Home Championship          4
4   CONMEBOL–UEFA Cup of Champions          1
5               Confederations Cup          9
6                     Copa América         17
7                   FIFA World Cup         16
8     FIFA World Cup qualification         39
9                         Gold Cup         18
10               NAFC Championship          3
11             Oceania Nations Cup         14
12                       UEFA Euro         14
13         UEFA Euro qualification         41
14             UEFA Nations League         14

5) Find which players are the top scorers of each tournament

In [16]:
# Merge the top scorers of each tournament with the maximum goals for that tournament
# 'player_and_tournament' contains players and their total goals for each tournament
# 'max_goals_tournament' contains the maximum number of goals scored in each tournament
# Do a left join so we can the 'max_goals' column for each tournament 
top_scorers_tournament = player_and_tournament.merge(max_goals_tournament, on='tournament', how='left')

# To find the top scorer of each tournament, 
# we can filter the dataframe to keep only rows where the player's 'goal_count' equals the 'max_goals' for that tournament
best_scorer_tournament = top_scorers_tournament[(top_scorers_tournament['goal_count']) == (top_scorers_tournament['max_goals'])]
best_scorer_tournament

tournament                   scorer  goal_count  \
33                      AFC Asian Cup                 Ali Daei          14   
1314           African Cup of Nations             Samuel Eto'o          18   
1601                       Baltic Cup          Ēriks Pētersons           9   
1611        British Home Championship              Geoff Hurst           4   
1623   CONMEBOL–UEFA Cup of Champions         Claudio Caniggia           1   
1624   CONMEBOL–UEFA Cup of Champions      Dominique Rocheteau           1   
1625   CONMEBOL–UEFA Cup of Champions               José Touré           1   
1626   CONMEBOL–UEFA Cup of Champions         Lautaro Martínez           1   
1627   CONMEBOL–UEFA Cup of Champions         Néstor Craviotto           1   
1628   CONMEBOL–UEFA Cup of Champions             Paulo Dybala           1   
1629   CONMEBOL–UEFA Cup of Champions           Ángel Di María           1   
1675               Confederations Cup        Cuauhtémoc Blanco           9   
1840               Confederations Cup               Ronaldinho           9   
2687                     Copa América  Norberto Doroteo Méndez          17   
2963                     Copa América                  Zizinho          17   
3932                   FIFA World Cup           Miroslav Klose          16   
5891     FIFA World Cup qualification              Carlos Ruiz          39   
14396                        Gold Cup           Landon Donovan          18   
14627               NAFC Championship        Luis de la Fuente           3   
14673             Oceania Nations Cup              Damian Mori          14   
14927                       UEFA Euro        Cristiano Ronaldo          14   
15885         UEFA Euro qualification        Cristiano Ronaldo          41   
18748             UEFA Nations League      Aleksandar Mitrović          14   

       max_goals  
33            14  
1314          18  
1601           9  
1611           4  
1623           1  
1624           1  
1625           1  
1626           1  
1627           1  
1628           1  
1629           1  
1675           9  
1840           9  
2687          17  
2963          17  
3932          16  
5891          39  
14396         18  
14627          3  
14673         14  
14927         14  
15885         41  
18748         14

6) Now we have the top scorers ("best_scorer_tournament") and the total number of goals ("tournament_goals") for each tournament in separate sets. We can merge the tournament goals with best_scorer to do the future calculations on one dataframe
    


In [17]:
best_scorer_tournament = best_scorer_tournament.merge(tournament_goals, on='tournament' )
best_scorer_tournament

tournament                   scorer  goal_count  \
0                    AFC Asian Cup                 Ali Daei          14   
1           African Cup of Nations             Samuel Eto'o          18   
2                       Baltic Cup          Ēriks Pētersons           9   
3        British Home Championship              Geoff Hurst           4   
4   CONMEBOL–UEFA Cup of Champions         Claudio Caniggia           1   
5   CONMEBOL–UEFA Cup of Champions      Dominique Rocheteau           1   
6   CONMEBOL–UEFA Cup of Champions               José Touré           1   
7   CONMEBOL–UEFA Cup of Champions         Lautaro Martínez           1   
8   CONMEBOL–UEFA Cup of Champions         Néstor Craviotto           1   
9   CONMEBOL–UEFA Cup of Champions             Paulo Dybala           1   
10  CONMEBOL–UEFA Cup of Champions           Ángel Di María           1   
11              Confederations Cup        Cuauhtémoc Blanco           9   
12              Confederations Cup               Ronaldinho           9   
13                    Copa América  Norberto Doroteo Méndez          17   
14                    Copa América                  Zizinho          17   
15                  FIFA World Cup           Miroslav Klose          16   
16    FIFA World Cup qualification              Carlos Ruiz          39   
17                        Gold Cup           Landon Donovan          18   
18               NAFC Championship        Luis de la Fuente           3   
19             Oceania Nations Cup              Damian Mori          14   
20                       UEFA Euro        Cristiano Ronaldo          14   
21         UEFA Euro qualification        Cristiano Ronaldo          41   
22             UEFA Nations League      Aleksandar Mitrović          14   

    max_goals  total_goals  
0          14          987  
1          18         1767  
2           9          252  
3           4         1811  
4           1            7  
5           1            7  
6           1            7  
7           1            7  
8           1            7  
9           1            7  
10          1            7  
11          9          423  
12          9          423  
13         17         2671  
14         17         2671  
15         16         2720  
16         39        23174  
17         18         1097  
18          3           26  
19         14          534  
20         14          829  
21         41         7967  
22         14         1153

7) Calculate the percentage that shows each top scorer's contribution to the total goals in the tournament (convert in percentage). And add the goal percentage in dataframe as column

In [18]:
best_scorer_tournament['goal_percentage'] = (best_scorer_tournament['goal_count'] / best_scorer_tournament['total_goals']) * 100
best_scorer_tournament

tournament                   scorer  goal_count  \
0                    AFC Asian Cup                 Ali Daei          14   
1           African Cup of Nations             Samuel Eto'o          18   
2                       Baltic Cup          Ēriks Pētersons           9   
3        British Home Championship              Geoff Hurst           4   
4   CONMEBOL–UEFA Cup of Champions         Claudio Caniggia           1   
5   CONMEBOL–UEFA Cup of Champions      Dominique Rocheteau           1   
6   CONMEBOL–UEFA Cup of Champions               José Touré           1   
7   CONMEBOL–UEFA Cup of Champions         Lautaro Martínez           1   
8   CONMEBOL–UEFA Cup of Champions         Néstor Craviotto           1   
9   CONMEBOL–UEFA Cup of Champions             Paulo Dybala           1   
10  CONMEBOL–UEFA Cup of Champions           Ángel Di María           1   
11              Confederations Cup        Cuauhtémoc Blanco           9   
12              Confederations Cup               Ronaldinho           9   
13                    Copa América  Norberto Doroteo Méndez          17   
14                    Copa América                  Zizinho          17   
15                  FIFA World Cup           Miroslav Klose          16   
16    FIFA World Cup qualification              Carlos Ruiz          39   
17                        Gold Cup           Landon Donovan          18   
18               NAFC Championship        Luis de la Fuente           3   
19             Oceania Nations Cup              Damian Mori          14   
20                       UEFA Euro        Cristiano Ronaldo          14   
21         UEFA Euro qualification        Cristiano Ronaldo          41   
22             UEFA Nations League      Aleksandar Mitrović          14   

    max_goals  total_goals  goal_percentage  
0          14          987         1.418440  
1          18         1767         1.018676  
2           9          252         3.571429  
3           4         1811         0.220872  
4           1            7        14.285714  
5           1            7        14.285714  
6           1            7        14.285714  
7           1            7        14.285714  
8           1            7        14.285714  
9           1            7        14.285714  
10          1            7        14.285714  
11          9          423         2.127660  
12          9          423         2.127660  
13         17         2671         0.636466  
14         17         2671         0.636466  
15         16         2720         0.588235  
16         39        23174         0.168292  
17         18         1097         1.640839  
18          3           26        11.538462  
19         14          534         2.621723  
20         14          829         1.688782  
21         41         7967         0.514623  
22         14         1153         1.214224

# ADDITIONAL 

Create and additional column that flags records with data quality issues

Resolve the identified quality issues

I didn't want to change the order of the tasks so I've put the data quality issues and solutions after the 5th question. This can be done before doing the calculations for the questions (I can also move the cells).

Before doing any operations, I have checked some potential data issues in each dataframe (I wrote the infos for each datasets before the first tasks solution, so I didn't need to write codes for them)
1) home_score and away_score in results dataframe are never negative numbers and the highest goals scored in one game is with 31 home goals which actually happened in Australia - American Samoa game. So there aren't any problems in score datas.
2)  minute in goalscorers dataframe starts from '1' (which is the earliest possible minute that a player can score in football) and ends with '122' so there aren't any problematic numbers. 
3) There are some missing 'minute', 'own_goal' and 'penalty' values in goalscorers datas.
4)  There are some missing 'first_shooter' values in shootouts datas.

1) Finding the duplicated datas

In [19]:
# Find the duplicated rows in 'results' dataframe and flag them
duplicates_results = results_df.duplicated(subset=['Game_ID'], keep=False)
results_df['data_quality_issues'] = duplicates_results
results_df[results_df['data_quality_issues'] == True]

date home_team      away_team  home_score  away_score tournament  \
9072   1974-02-17    Tahiti  New Caledonia           2           1   Friendly   
9073   1974-02-17    Tahiti  New Caledonia           1           2   Friendly   
10554  1977-10-22    Guyana       Barbados           2           0   Friendly   
10555  1977-10-22    Guyana       Barbados           0           0   Friendly   

             city           country  neutral                          Game_ID  \
9072      Papeete  French Polynesia    False  1974-02-17_Tahiti_New Caledonia   
9073      Papeete  French Polynesia    False  1974-02-17_Tahiti_New Caledonia   
10554      Linden            Guyana    False       1977-10-22_Guyana_Barbados   
10555  Georgetown            Guyana    False       1977-10-22_Guyana_Barbados   

       total_goals  data_quality_issues  
9072             3                 True  
9073             3                 True  
10554            2                 True  
10555            0                 True

In [20]:
# Fixing the duplicated rows in results dataframe
results_df = results_df.drop_duplicates(subset=['Game_ID'], keep='first')

# Unflag the duplicated rows after dropping the one of the copy's  
duplicates_results = results_df.duplicated(subset=['Game_ID'], keep=False)
results_df['data_quality_issues'] = duplicates_results

results_df[results_df['data_quality_issues'] == True]

Empty DataFrame
Columns: [date, home_team, away_team, home_score, away_score, tournament, city, country, neutral, Game_ID, total_goals, data_quality_issues]
Index: []

In [21]:
# Find the duplicated rows in shootouts and flag them 
duplicates_shootouts = shootout_df.duplicated(subset=['Game_ID','first_shooter'], keep=False)
shootout_df['data_quality_issues'] = duplicates_shootouts
shootout_df[shootout_df['data_quality_issues'] == True]

Empty DataFrame
Columns: [date, home_team, away_team, winner, first_shooter, Game_ID, data_quality_issues]
Index: []

In [22]:
# Find the duplicated rows in goalscorers and flag them 
duplicates_goalscorers = goalscorers_df.duplicated(subset=['Game_ID', 'minute', 'scorer'], keep=False)
goalscorers_df['data_quality_issues'] = duplicates_goalscorers
# In goalscorers dataset, there are some 'NaN' minute values (which is another quality issue), 
# One player can score more than one goal in the same game but cannot score more than two goals in one game in same minutes.
# So I've filtered NaN minutes in dataset so it only shows the players who scored in same minute in the same game
goalscorers_df[(goalscorers_df['data_quality_issues'] == True) & (goalscorers_df['minute'].notna())]

date     home_team away_team         team               scorer  \
8757   1981-06-06          Fiji    Taiwan         Fiji            Ratu Jone   
8758   1981-06-06          Fiji    Taiwan         Fiji            Ratu Jone   
20313  2001-04-30          Oman      Laos         Oman       Hani Al-Dhabit   
20314  2001-04-30          Oman      Laos         Oman       Hani Al-Dhabit   
21447  2002-01-30  Burkina Faso     Ghana        Ghana         Isaac Boakye   
...           ...           ...       ...          ...                  ...   
42607  2023-06-18   Netherlands     Italy  Netherlands      Steven Bergwijn   
42608  2023-06-18   Netherlands     Italy        Italy      Federico Chiesa   
42609  2023-06-18   Netherlands     Italy        Italy      Federico Chiesa   
42610  2023-06-18   Netherlands     Italy  Netherlands  Georginio Wijnaldum   
42611  2023-06-18   Netherlands     Italy  Netherlands  Georginio Wijnaldum   

       minute own_goal penalty                        Game_ID  \
8757     55.0    False   False         1981-06-06_Fiji_Taiwan   
8758     55.0    False   False         1981-06-06_Fiji_Taiwan   
20313    45.0    False   False           2001-04-30_Oman_Laos   
20314    45.0    False   False           2001-04-30_Oman_Laos   
21447    90.0    False   False  2002-01-30_Burkina Faso_Ghana   
...       ...      ...     ...                            ...   
42607    68.0    False   False   2023-06-18_Netherlands_Italy   
42608    72.0    False   False   2023-06-18_Netherlands_Italy   
42609    72.0    False   False   2023-06-18_Netherlands_Italy   
42610    89.0    False   False   2023-06-18_Netherlands_Italy   
42611    89.0    False   False   2023-06-18_Netherlands_Italy   

       data_quality_issues  
8757                  True  
8758                  True  
20313                 True  
20314                 True  
21447                 True  
...                    ...  
42607                 True  
42608                 True  
42609                 True  
42610                 True  
42611                 True  

[96 rows x 10 columns]

In [23]:
# Fixing the duplicated rows in goalscorers dataframe
goalscorers_df = goalscorers_df.drop_duplicates(subset=['Game_ID', 'minute', 'scorer'], keep='first')

# Unflag the duplicated rows after dropping the one of the copy's  
duplicates_goalscorers = goalscorers_df.duplicated(subset=['Game_ID', 'minute', 'scorer'], keep=False)
goalscorers_df['data_quality_issues'] = duplicates_goalscorers


goalscorers_df[(goalscorers_df['data_quality_issues'] == True) & (goalscorers_df['minute'].notna())]


Empty DataFrame
Columns: [date, home_team, away_team, team, scorer, minute, own_goal, penalty, Game_ID, data_quality_issues]
Index: []